lib/cchain/provider/attestations.go

```diff
@@ -0,0 +1,350 @@
+package provider
+
+import (
+	"context"
+
+	atypes "github.com/omni-network/omni/halo/attest/types"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/tracer"
+	"github.com/omni-network/omni/lib/umath"
+	"github.com/omni-network/omni/lib/xchain"
+
+	"github.com/cosmos/cosmos-sdk/client/grpc/cmtservice"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type attestations struct {
+	query      atypes.QueryClient
+	service    cmtservice.ServiceClient
+	chainNamer func(xchain.ChainVersion) string
+}
+
+func (a attestations) Find(
+	ctx context.Context,
+	chainVer xchain.ChainVersion,
+	fromOffset uint64,
+	cachedHeight uint64,
+) ([]xchain.Attestation, uint64, error) {
+	log.Debug(ctx, "[atts] find", 33, "cached", cachedHeight, "from", fromOffset)
+
+	const endpoint = "fetch_attestations"
+	defer latency(endpoint)()
+
+	ctx, span := tracer.Start(ctx, spanName(endpoint))
+	defer span.End()
+
+	chainName := a.chainNamer(chainVer)
+
+	// try fetching from latest height
+	atts, ok, err := a.fromAtHeight(ctx, chainVer, fromOffset, 0)
+	if err != nil {
+		incQueryErr(endpoint)
+		return nil, 0, errors.Wrap(err, "abci query attestations-from")
+	} else if ok {
+		debugLogAtts(ctx, "[atts] find [49]", atts)
+		fetchStepsMetrics(chainName, 0, 0, 0)
+
+		return atts, 0, nil
+	}
+
+	earliestAttestationAtLatestHeight, ok, err := a.earliestAtHeight(ctx, chainVer, 0)
+	if err != nil {
+		incQueryErr(endpoint)
+
+		return nil, 0, errors.Wrap(err, "abci query earliest-attestation-in-state")
+	}
+
+	// Either no attestations have happened yet, or the queried fromOffset is in the "future"
+	// Caller has to wait and retry in both cases
+	if !ok || earliestAttestationAtLatestHeight.AttestOffset < fromOffset {
+		// First attestation hasn't happened yet, return empty
+		return []xchain.Attestation{}, 0, nil
+	}
+
+	latestBlockResp, err := a.service.GetLatestBlock(ctx, &cmtservice.GetLatestBlockRequest{})
+	if err != nil {
+		return []xchain.Attestation{}, 0, errors.Wrap(err, "query latest block")
+	}
+	latestHeight := uint64(latestBlockResp.SdkBlock.Header.Height)
+
+	var startHeight, endHeight uint64
+	// if we are performing search for the first time and no previous height was cached do a full search
+	if cachedHeight == 0 {
+		startHeight, endHeight, err = a.lookbackRange(ctx, chainVer, fromOffset, latestHeight)
+		if err != nil {
+			incQueryErr(endpoint)
+			return nil, 0, errors.Wrap(err, "lookback search")
+		}
+	} else {
+		startHeight = cachedHeight + 1 // we move to the next height
```

Comment:
not sure we can skip cached height, it might still contain the attestation we are looking for, best start from this point (inclusive)